bluetooth: rfcomm: fix issue of sending buf invalid

lylezhu2012 opened the pull request:
There is a change (commit no.: 93d0eac834) that if the `ref` of sending buf is not 1, the error code `-EINVAL` will be returned from bt_conn_send_cb.

It causes the RFCOMM functionality cannot work properly.

Remove the ref operation from the buf to be sent to fix the issue.

Due to the sending buf cannot be referred by sending layer.
It is unsafe to use `buf` identification for a transmission, because the buf may have been newly transmitted when the
sent callback is triggered.

Now instead, when the send completion callback is received, the upper layer is notified that a transfer is completed.
If multiple bufs are sent at the same time, there is no guarantee which buf is completed when the sent callback triggered. Therefore, it is recommended that the caller transfers the next data block after the previous transfer is completed.

HFP AG: Due to the sent callback of RFCOMM is changed, the sending buf need to be primed waiting for the previous one to be completed. Add a worker for this purpose.

HFP Unit: Due to the parameter `buf` has been removed by rfcomm, update the prototype of channel sent callback hfp_hf_sent.

lylezhu2012 committed:
bluetooth: rfcomm: fix issue of sending buf invalid

There is a change (commit no.: 93d0eac834) that if
the `ref` of sending buf is not 1, the error code
`-EINVAL` will be returned from bt_conn_send_cb.

It causes the RFCOMM functionality cannot work
properly.

Remove the ref operation from the buf to be sent
to fix the issue.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
bluetooth: rfcomm: remove tx meta

Due to the sending buf cannot be referred
by sending layer.
It is unsafe to use `buf` identification
for a transmission, because the buf may
have been newly transmitted when the
sent callback is triggered.

Now instead, when the send completion
callback is received, the upper layer
is notified that a transfer is completed.
If multiple bufs are sent at the same
time, there is no guarantee which buf
is completed when the sent callback
triggered. Therefore, it is recommended
that the caller transfers the next data
block after the previous transfer is
completed.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: HFP_AG: Optimize the TX process

Due to the sent callback of RFCOMM is changed, the
sending buf need to be primed waiting for the
previous one to be completed. Add a worker for
this purpose.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
bluetooth: hfp_hf: update channel sent callback prototype

Due to the parameter `buf` has been removed by rfcomm,
update the prototype of channel sent callback hfp_hf_sent.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jhedberg commented:
@lylezhu2012 seems like the issue was a reference leak, i.e. `bt_rfcomm_dlc_send()` was always supposed to move the reference rather than taking a new one. I think ideally it should be taking a `struct net_buf **buf` as a parameter and setting the original pointer to NULL to make the reference move clear. Btw, this also relates to #72798.

At a bare minimum I think it'd be important to accurately describe the bug that you're fixing. Now it sounds like you're working around an issue introduced by commit 93d0eac8345a60497abf0c89b5f26c41a59ea251, but that's actually not the case - the reference leak existed even before that commit.

jhedberg reviewed (changes requested):
@lylezhu2012 also, the documentation for `bt_rfcomm_dlc_send() is missing the information bout the API moving the reference. I think that's a potential source of future bugs, so please add the missing information to the documentation as well.

lylezhu2012 commented:
> @lylezhu2012 also, the documentation for `bt_rfcomm_dlc_send() is missing the information bout the API moving the reference. I think that's a potential source of future bugs, so please add the missing information to the documentation as well.

OK. But I plan to revert it in this PR. Do not update the reference of buf.

jhedberg commented on subsys/bluetooth/host/classic/hfp_ag.c:
```diff
@@ -266,7 +268,60 @@ static int hfp_ag_next_step(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *us
+		return;
+	}
+
+	hfp_ag_lock(ag);
+	node = sys_slist_peek_head(&ag->tx_pending);
+	hfp_ag_unlock(ag);
+	if (!node) {
+		LOG_DBG("No pending tx");
+		return;
+	}
+
+	tx = CONTAINER_OF(node, struct bt_ag_tx, node);
```
Since you're not holding the lock anymore at this point, doesn't it mean that the list could have changed and the data invalidated. It seems to me like you should be holding the lock until you're completely done with accessing the `ag` pointer in this function.

lylezhu2012 replied:
Updated.

lylezhu2012 commented:
@jhedberg , @sjanc , @alwa-nordic , Could you please take a look at it.

dleach02 commented:
@jhedberg can you review to determine if your change request is satisfied.

jhedberg commented on subsys/bluetooth/host/classic/hfp_ag.c:
```diff
@@ -266,7 +268,59 @@ static int hfp_ag_next_step(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *us
+	struct bt_ag_tx *tx;
+	bt_hfp_state_t state;
+
+	hfp_ag_lock(ag);
+	state = ag->state;
+	hfp_ag_unlock(ag);
+	if ((state == BT_HFP_DISCONNECTED) || (state == BT_HFP_DISCONNECTING)) {
+		LOG_ERR("AG %p is not connected", ag);
+		return;
+	}
+
+	hfp_ag_lock(ag);
+	node = sys_slist_peek_head(&ag->tx_pending);
+	if (!node) {
+		hfp_ag_unlock(ag);
+		LOG_DBG("No pending tx");
+		return;
+	}
```
Why do you have a different pattern for locking and unlocking with the first and second branch? In the first one you lock before the branch and in the second one in the branch.

Generally, when you have a function which holds a lock for a large part of the function, the typical pattern is:

```c
static void my_func(struct bt_hfp_ag *ag)
{
	int err;

	hfp_ag_lock(ag);
	err = foo(ag);
	if (err) {
		LOG_ERR("...");
		goto unlock;
	}	

	err = bar(ag);
	if (err) {
		LOG_ERR("...");
		goto unlock;
	}	

	...

unlock:
	hfp_ag_unlock(ag);
	return err;
}
```

The exception to the above is if the function does callback or anything that might introduce re-entrancy to try to acquire the lock a second time. In such cases you need to release and re-acquire the lock, or use a different lock type which permits the same thread to take the lock multiple times.

lylezhu2012 replied:
Updated

lylezhu2012 committed:
Bluetooth: HFP_AG: Optimize lock/unlock of the SCO creating

When testing https://github.com/zephyrproject-rtos/zephyr/pull/72090/,
there is an issue found.

The change in the previous commit is to put all data sending
operations into the work queue context, and lock the current
AG before sending data.

And in change of #72090, the HCI TX thread is removed. All
sending sequence are happened in work queue context.

There is a possible problem when AG creates a SCO connection
by calling the function bt_conn_create_sco. Before the
function bt_conn_create_sco is called, AG will be locked to
avoid creating repeated SCO connection.
And the execution of the function bt_conn_create_sco
depends on the work queue. Because the HCI command of
function bt_conn_create_sco is sent in work queue context.

In the normal case, there is not any issue.
But there is a case that when the function
bt_conn_create_sco is being executed, there is a pending AG
TX waiting to be executed.

Once the work queue starts executing the handler, the AG TX
handler is executed first. Since the lock has been acquired
by other threads, the AG TX handler cannot acquire the lock.
As a result, the SCO connection creation fails.

Remove the AG lock from SCO creating. Instead, use a flag
to mark whether a SCO connection is be created.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: HFP_AG: fix building warning

Give `err` a initialization value 0.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented:
When testing https://github.com/zephyrproject-rtos/zephyr/pull/72090/, there is an issue found.

The change in the previous commit is to put all data sending operations into the work queue context, and lock the current
AG before sending data.

And in change of #72090, the HCI TX thread is removed. All sending sequence are happened in work queue context.

There is a possible problem when AG creates a SCO connection by calling the function bt_conn_create_sco. Before the
function bt_conn_create_sco is called, AG will be locked to avoid creating repeated SCO connection.
And the execution of the function bt_conn_create_sco depends on the work queue. Because the HCI command of
function bt_conn_create_sco is sent in work queue context.

In the normal case, there is not any issue. 

But there is a case that when the function bt_conn_create_sco is being executed, there is a pending AG TX waiting to be executed.

Once the work queue starts executing the handler, the AG TX handler is executed first. Since the lock has been acquired
by other threads, the AG TX handler cannot acquire the lock. As a result, the SCO connection creation fails.

Remove the AG lock from SCO creating. Instead, use a flag to mark whether a SCO connection is be created.

lylezhu2012 commented:
@dleach02 , @carlescufi , now there is only one reviewer approves the PR. Is there anyone else who can help review this PR? Currently, handsfree and handsfree_ag based on RFCOMM cannot work properly.

lylezhu2012 commented:
> @dleach02 , @carlescufi , now there is only one reviewer approves the PR. Is there anyone else who can help review this PR? Currently, handsfree and handsfree_ag based on RFCOMM cannot work properly.
